workos/directory_sync.py

```diff
@@ -297,7 +297,229 @@ def list_directories(
             **ListPage[Directory](**response).model_dump(),
         )
 
-    def delete_directory(self, directory: str):
+    def delete_directory(self, directory: str) -> None:
+        """Delete one existing Directory.
+
+        Args:
+            directory (str): The ID of the directory to be deleted. (Required)
+
+        Returns:
+            None
+        """
+        self._http_client.request(
+            "directories/{directory}".format(directory=directory),
+            method=REQUEST_METHOD_DELETE,
+            token=workos.api_key,
+        )
+
+
+class AsyncDirectorySync(DirectorySyncModule):
+    """Offers methods through the WorkOS Directory Sync service."""
+
+    _http_client: AsyncHTTPClient
+
+    @validate_settings(DIRECTORY_SYNC_MODULE)
+    def __init__(self, http_client: AsyncHTTPClient):
+        self._http_client = http_client
+
+    async def list_users(
+        self,
+        directory: Optional[str] = None,
+        group: Optional[str] = None,
+        limit: int = RESPONSE_LIMIT,
+        before: Optional[str] = None,
+        after: Optional[str] = None,
+        order: PaginationOrder = "desc",
+    ) -> AsyncWorkOsListResource[DirectoryUser, DirectoryUserListFilters]:
+        """Gets a list of provisioned Users for a Directory.
+
+        Note, either 'directory' or 'group' must be provided.
+
+        Args:
+            directory (str): Directory unique identifier.
+            group (str): Directory Group unique identifier.
+            limit (int): Maximum number of records to return.
+            before (str): Pagination cursor to receive records before a provided Directory ID.
+            after (str): Pagination cursor to receive records after a provided Directory ID.
+            order (Order): Sort records in either ascending or descending order by created_at timestamp.
+
+        Returns:
+            dict: Directory Users response from WorkOS.
+        """
+
+        list_params = {
+            "limit": limit,
+            "before": before,
+            "after": after,
+            "order": order,
+        }
+
+        if group is not None:
+            list_params["group"] = group
+        if directory is not None:
+            list_params["directory"] = directory
+
+        response = await self._http_client.request(
+            "directory_users",
+            method=REQUEST_METHOD_GET,
+            params=list_params,
+            token=workos.api_key,
+        )
+
+        return AsyncWorkOsListResource(
+            list_method=self.list_users,
+            list_args=list_params,
+            **ListPage[DirectoryUser](**response).model_dump(),
+        )
```

Comment:
This PR lacks deduplication of code like this. Might be a good thing for me to take a stab at either before or after I work on connections.